presto-parser/src/main/java/com/facebook/presto/sql/tree/ShowStats.java

```diff
@@ -13,32 +13,50 @@
  */
 package com.facebook.presto.sql.tree;
 
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 
+import static com.facebook.presto.sql.QueryUtil.simpleQuery;
 import static com.google.common.base.MoreObjects.toStringHelper;
-import static java.util.Collections.emptyList;
 
 public class ShowStats
         extends Statement
 {
-    private final QualifiedName table;
+    private final Query query;
 
+    @VisibleForTesting
     public ShowStats(QualifiedName table)
     {
         this(Optional.empty(), table);
     }
 
-    public ShowStats(NodeLocation location, QualifiedName table)
+    @VisibleForTesting
+    public ShowStats(Query query)
+    {
+        this(Optional.empty(), query);
+    }
+
+    public ShowStats(Optional<NodeLocation> location, QualifiedName table)
     {
-        this(Optional.of(location), table);
+        this(location, createFakeQuery(location, table));
     }
 
-    private ShowStats(Optional<NodeLocation> location, QualifiedName table)
+    public ShowStats(Optional<NodeLocation> location, Query query)
     {
         super(location);
-        this.table = table;
+        this.query = query;
+    }
+
+    private static Query createFakeQuery(Optional<NodeLocation> location, QualifiedName name)
```

Comment:
>  I agree that it's added value, but on the other hand this would additionally complicate ShowStatsRewrite as we would have to cover both cases separately there. 

ShowStatsRewrite is what should be performing all semantic checks, including, checking for what's supported vs not.

> TLDR: This change can be done, but my personal opinion is that it's only moving complexity from one point to the other and for that reason it's probably not worth additional work.

I disagree. The complexity needs to exist somewhere (as long as we don't support arbitrary query shapes), but the parser is the wrong place for it. As the rewrite becomes more powerful over time, it can relax the checks and remove unnecessary code without having to touch the parser or AST any further.